frame/merkle-mountain-range/src/lib.rs

```diff
@@ -318,38 +318,81 @@ impl<T: Config<I>, I: 'static> Pallet<T, I> {
 			.saturating_add(leaf_index.saturated_into())
 	}
 
-	/// Generate a MMR proof for the given `leaf_indices`.
+	/// Convert a `block_num` into a leaf index.
+	fn block_num_to_leaf_index(block_num: T::BlockNumber) -> Result<LeafIndex, primitives::Error>
+	where
+		T: frame_system::Config,
+	{
+		// leaf_indx = block_num - (current_block_num - leaves_count) - 1;
+		let leaves_count = Self::mmr_leaves().saturated_into::<u64>();
+		let current_block_num = <frame_system::Pallet<T>>::block_number().saturated_into::<u64>();
+		let diff = current_block_num.saturating_sub(leaves_count.into());
+
+		let block_num_as_u64 = block_num.saturated_into::<u64>();
+
+		if block_num_as_u64 <= diff {
+			return Err(
+				primitives::Error::BlockNumToLeafIndex.log_debug("The block_number is incorrect.")
+			)
+		}
+
+		let leaf_index = (block_num_as_u64.saturating_sub(diff).saturating_sub(1u32.into()))
+			.saturated_into::<LeafIndex>();
+		Ok(leaf_index)
+	}
+
+	/// Generate a MMR proof for the given `block_numbers`.
 	///
 	/// Note this method can only be used from an off-chain context
 	/// (Offchain Worker or Runtime API call), since it requires
 	/// all the leaves to be present.
 	/// It may return an error or panic if used incorrectly.
 	pub fn generate_batch_proof(
-		leaf_indices: Vec<LeafIndex>,
+		block_numbers: Vec<T::BlockNumber>,
 	) -> Result<
 		(Vec<LeafOf<T, I>>, primitives::BatchProof<<T as Config<I>>::Hash>),
 		primitives::Error,
 	> {
-		Self::generate_historical_batch_proof(leaf_indices, Self::mmr_leaves())
+		Self::generate_historical_batch_proof(
+			block_numbers,
+			Self::leaf_index_to_parent_block_num(Self::mmr_leaves(), Self::mmr_leaves()),
+		)
 	}
 
-	/// Generate a MMR proof for the given `leaf_indices` for the MMR of `leaves_count` size.
+	/// Generate a MMR proof for the given `block_numbers` for the MMR of `leaves_count` size.
 	///
 	/// Note this method can only be used from an off-chain context
 	/// (Offchain Worker or Runtime API call), since it requires
 	/// all the leaves to be present.
 	/// It may return an error or panic if used incorrectly.
 	pub fn generate_historical_batch_proof(
-		leaf_indices: Vec<LeafIndex>,
-		leaves_count: LeafIndex,
+		block_numbers: Vec<T::BlockNumber>,
+		best_known_block_number: T::BlockNumber,
 	) -> Result<
 		(Vec<LeafOf<T, I>>, primitives::BatchProof<<T as Config<I>>::Hash>),
 		primitives::Error,
 	> {
+		let now = frame_system::Pallet::<T>::block_number().saturated_into::<u64>();
+		let leaves = Self::mmr_leaves().saturated_into::<u64>();
+		let best_block_as_u64 = best_known_block_number.saturated_into::<u64>();
+
+		let leaves_count = leaves
+			.saturating_sub(now)
+			.saturating_add(best_block_as_u64)
+			.saturated_into::<LeafIndex>();
```

Comment:
actually, why not just replace this with
```suggestion
		let leaves_count =
			Self::block_num_to_leaf_index(best_known_block_number)?.saturating_add(1);
```
then you also check that your historic best block is not prior to pallet activation.
and you can simplify further by refactoring `block_num_to_leaf_index` to `block_num_to_leaf_count`, unless you want to keep an exact inverse to `leaf_index_to_block_num` around.

Reply:
Not sure I understood this the best, but we still need the `block_num_to_leaf_index` function, there could be a new `block_num_to_leaf_count` function, but I don't see why we wouldn't just do this
```rs
let leaves_count = Self::block_num_to_leaf_index(best_known_block_number)?.saturating_add(1);
```
as you suggested. 
Or would this be used somewhere else in the code, in that case, the new function makes sense.
Sorry if I didn't understand you correctly.

Reply:
> we still need the `block_num_to_leaf_index` function

thanks: I oversaw that you're still using it here
https://github.com/paritytech/substrate/blob/4538c0832e81d1bff5df9ff821c5515326bcc54e/frame/merkle-mountain-range/src/lib.rs#L385

so mistakenly thought this was the only consumer, and could tailor interface to it. So makes more sense to keep like original suggestion, as you've done :+1: